src/ray/object_manager/notification/object_store_notification_manager.h

```diff
@@ -0,0 +1,95 @@
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RAY_OBJECT_STORE_NOTIFICATION_MANAGER_H
+#define RAY_OBJECT_STORE_NOTIFICATION_MANAGER_H
+
+#include <iostream>
+#include <memory>
+#include <vector>
+
+#include <boost/asio.hpp>
+
+#include "ray/common/id.h"
+#include "ray/common/status.h"
+#include "ray/object_manager/format/object_manager_generated.h"
+
+namespace ray {
+
+/// \class ObjectStoreNotificationManager
+///
+/// Encapsulates notification handling from the object store.
+class ObjectStoreNotificationManager {
+ public:
+  ObjectStoreNotificationManager(boost::asio::io_service &io_service)
+      : main_service_(&io_service), num_adds_processed_(0), num_removes_processed_(0) {}
+  virtual ~ObjectStoreNotificationManager() {}
+  /// Subscribe to notifications of objects added to local store.
+  /// Upon subscribing, the callback will be invoked for all objects that
+  /// already exist in the local store
+  ///
+  /// \param callback A callback expecting an ObjectID.
+  void SubscribeObjAdded(
+      std::function<void(const object_manager::protocol::ObjectInfoT &)> callback) {
+    add_handlers_.push_back(std::move(callback));
+  }
+
+  /// Subscribe to notifications of objects deleted from local store.
+  ///
+  /// \param callback A callback expecting an ObjectID.
+  void SubscribeObjDeleted(std::function<void(const ray::ObjectID &)> callback) {
+    rem_handlers_.push_back(std::move(callback));
+  }
+
+  /// Support for rebroadcasting object add/rem events.
+  void ProcessStoreAdd(const object_manager::protocol::ObjectInfoT &object_info) {
+    // TODO(suquark): Use strand in boost asio to enforce sequential execution.
+    for (auto &handler : add_handlers_) {
+      main_service_->post([handler, object_info]() { handler(object_info); });
+    }
+    num_adds_processed_++;
+  }
+
+  void ProcessStoreRemove(const ObjectID &object_id) {
+    for (auto &handler : rem_handlers_) {
+      main_service_->post([handler, object_id]() { handler(object_id); });
+    }
+    num_removes_processed_++;
+  }
+
+  /// Returns debug string for class.
+  ///
+  /// \return string.
+  std::string DebugString() const {
+    std::stringstream result;
+    result << "ObjectStoreNotificationManager:";
+    result << "\n- num adds processed: " << num_adds_processed_;
+    result << "\n- num removes processed: " << num_removes_processed_;
+    return result.str();
+  }
+
+ private:
+  /// Weak reference to main service. We ensure this object is destroyed before
+  /// main_service_ is stopped.
+  boost::asio::io_service *main_service_;
```

Comment:
Would it be better to use a shared_ptr just in case?

Reply:
It comes from the original code. do you think it would be better to use shared_ptr? I think shared_ptr is useless here since it is not `shared` by others because the incoming `io_service` is not a shared_ptr.